Implemented Bitrix24 provider

Juriyx opened the pull request.

Commit:
Implemented Bitrix24 provider

Juriyx commented:
@martincostello, I tested the provider

xperiandri commented:
@martincostello, I tested the provider

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationConstants.cs:
```diff
@@ -0,0 +1,258 @@
+            public const string MiddleName = "second_name";
+            public const string City = "personal_city";
+            public const string Photo = "personal_photo";
+            public const string ICQ = "personal_icq";
```
```suggestion
            public const string Icq = "personal_icq";
```

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationConstants.cs:
```diff
@@ -0,0 +1,258 @@
+        public static class Scopes
+        {
+            /// <summary>
+            /// <a href="https://dev.1c-bitrix.ru/rest_help/bizproc/index.php" target="_blank">Бизнес-процессы</a>
```
Please make these comments written sentences (not just a link) so they're useful in Intellisense in Visual Studio for example. Also please use English.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationConstants.cs:
```diff
@@ -0,0 +1,258 @@
+﻿/*
```
There's _a lot_ of constants in this file. Do they really all need to be built in?

Juriyx replied:
Well, they just represent available fields of user info endpoint. Do you think that they must not be there?

martincostello replied:
The library doesn't _have_ to have _every_ possible thing built in, just the common defaults. Particularly as they're just strings and there's nothing stopping any integrator passing in literal strings, or making their own string constants classes with the extra ones they use/need in it.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationDefaults.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        public const string CallbackPath = "/signin-bitrix24";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpoint = "https://{0}/oauth/authorize/";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpoint = "https://{0}/oauth/token/";
+
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.UserInformationEndpoint"/>.
+        /// </summary>
+        public const string UserInformationEndpoint = "https://{0}/rest/user.current.json";
```
Please refactor to use the same pattern used by the Okta and Moodle providers for handling different subdomains using an `IPostConfigureOptions<T>` implementation.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationHandler.cs:
```diff
@@ -0,0 +1,83 @@
+        {
+            var authorizationHeader = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+            Backchannel.DefaultRequestHeaders.Authorization = authorizationHeader;
+            using var response = await Backchannel.GetAsync(Options.UserInformationEndpoint, Context.RequestAborted);
```
```suggestion

            using var response = await Backchannel.GetAsync(Options.UserInformationEndpoint, Context.RequestAborted);
```

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationHandler.cs:
```diff
@@ -0,0 +1,83 @@
+                throw new HttpRequestException($"An error occurred when retrieving Bitrix24 user information ({response.StatusCode}). Please check if the authentication information is correct.");
+            }
+
+            using (var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted)))
+            {
+                var context = new OAuthCreatingTicketContext(new ClaimsPrincipal(identity), properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement.GetProperty("result"));
+                context.RunClaimActions();
+                await Events.CreatingTicket(context);
+
+                return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+            }
```
```suggestion
            var payload = JsonDocument.Parse(await response.Content.ReadAsStringAsync(Context.RequestAborted));

            var principal = new ClaimsPrincipal(identity);
            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
            context.RunClaimActions(payload.RootElement.GetProperty("result"));
            await Events.CreatingTicket(context);

            return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
```

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationHandler.cs:
```diff
@@ -0,0 +1,83 @@
+            using var response = await Backchannel.GetAsync(Options.UserInformationEndpoint, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new HttpRequestException($"An error occurred when retrieving Bitrix24 user information ({response.StatusCode}). Please check if the authentication information is correct.");
```
```suggestion
                                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
                                "returned a {Status} response with the following payload: {Headers} {Body}.",
                                /* Status: */ response.StatusCode,
                                /* Headers: */ response.Headers.ToString(),
                                /* Body: */ await response.Content.ReadAsStringAsync(Context.RequestAborted));

                throw new HttpRequestException("An error occurred while retrieving the user profile from Bitrix24.");
```

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationOptions.cs:
```diff
@@ -0,0 +1,85 @@
+    {
+        private string? tenantHostName;
+
+        public string? TenantHostName
+        {
+            get
+            {
+                return tenantHostName;
+            }
+
+            set
+            {
+                tenantHostName = value;
+                AuthorizationEndpoint = string.Format(CultureInfo.InvariantCulture, Bitrix24AuthenticationDefaults.AuthorizationEndpoint, value);
+                TokenEndpoint = string.Format(CultureInfo.InvariantCulture, Bitrix24AuthenticationDefaults.TokenEndpoint, value);
+                UserInformationEndpoint = string.Format(CultureInfo.InvariantCulture, Bitrix24AuthenticationDefaults.UserInformationEndpoint, value);
+            }
+        }
```
As noted earlier, refactor this to use the pattern the Okta provider (and others) use for varying host names.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationOptions.cs:
```diff
@@ -0,0 +1,85 @@
+            TokenEndpoint = Bitrix24AuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = Bitrix24AuthenticationDefaults.UserInformationEndpoint;
+
+            ClaimActions.MapJsonKey(ClaimTypes.NameIdentifier, UserFields.Id);
```
Mapping everything is going to create a very large authentication cookie storing all of these claims.

Are they really all needed by default?

The normal pattern is to only automatically configure the basic/required claims, and users can map additional ones themselves in their individual configuration.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/Bitrix24/Bitrix24Tests.cs:
```diff
@@ -0,0 +1,52 @@
+            });
+        }
+
+        [Theory]
```
All claims mapped by the library should have a test case.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/OAuthTests`1.cs:
```diff
@@ -41,12 +41,14 @@ protected OAuthTests()
                 .ThrowsOnMissingRegistration()
                 .RegisterBundle(Path.Combine(BundleName, "bundle.json"));
 
+#pragma warning disable DF0022 // Marks indisposed objects assinged to a property, originated in an object creation.
```
What tooling raises this warning?

Juriyx replied:
https://github.com/BADF00D/DisposableFixer

martincostello replied:
Given we don't use this analyzer in the repo source code and this is just something specific to your IDE setup, I'd rather it be removed.

xperiandri replied:
But as long as you treat warnings as errors we can't build the project without that.

xperiandri replied:
What about adding this analyzer?

martincostello replied:
I'd prefer to just fix the problem it's identifying (even though it's not an issue in practice), rather than add an extra analyser just because you happen to have it installed in your personal Visual Studio installation.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/OAuthTests`1.cs:
```diff
@@ -130,7 +132,10 @@ protected WebApplicationFactory<Program> CreateTestServer(Action<IServiceCollect
         protected HttpClient CreateBackchannel(AuthenticationBuilder builder)
-            => builder.Services.BuildServiceProvider().GetRequiredService<IHttpClientFactory>().CreateClient();
+        {
+            using var serviceProvider = builder.Services.BuildServiceProvider();
```
Why is a scope suddenly needed here?

Juriyx replied:
https://github.com/BADF00D/DisposableFixer suggested that there is indisposed object

Commit:
Implemented suggestions by @martincostello

Commit:
Added vertical bar in the last column between "Documentation" and "Claims"

Commit:
Fixed other providers

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationConstants.cs:
```diff
@@ -0,0 +1,272 @@
+        public static class Scopes
+        {
+            /// <summary>
+            /// Workflows (<a href="https://training.bitrix24.com/rest_help/workflows/index.php">documentation</a>)
```
It's OK for the links to be to something in Russian, it was just the _code_ comments I was referring to needing to be in English. We don't really want to have two sets of comments to maintain for each property, with one of them being in a language the repo maintainers don't speak.

xperiandri replied:
The point we decided to go that way is that Bitrix24 is built in Russia, so that it most probably will be used by Russia speaking developer.
And as you can see with Workflows the scope has totally different name in Russian. If you strongly believe that it is not reasonable we of course can leave the English only name but 2 links.

martincostello replied:
Most providers added to this repo end up never being revisited by the original authors, and then being looked after by the maintainers such as myself.

To make things easier for us to maintain long term, English comments with links to Russian documentation for further reading is preferred.

martincostello commented on README.md:
```diff
@@ -125,6 +125,7 @@ If a provider you're looking for does not exist, consider making a PR to add one
 | BattleNet | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.BattleNet?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.BattleNet/ "Download AspNet.Security.OAuth.BattleNet from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.BattleNet?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/packag ...
 | Beam (Mixer) | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.Beam?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.Beam/ "Download AspNet.Security.OAuth.Beam from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.Beam?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/package/nuget/AspNet.Se ...
 | Bitbucket | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.Bitbucket?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.Bitbucket/ "Download AspNet.Security.OAuth.Bitbucket from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.Bitbucket?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/packag ...
+| Bitrix24 | [![NuGet](https://buildstats.info/nuget/AspNet.Security.OAuth.Bitrix24?includePreReleases=false)](https://www.nuget.org/packages/AspNet.Security.OAuth.Bitrix24/ "Download AspNet.Security.OAuth.Bitrix24 from NuGet.org") | [![MyGet](https://buildstats.info/myget/aspnet-contrib/AspNet.Security.OAuth.Bitrix24?includePreReleases=true)](https://www.myget.org/feed/aspnet-contrib/package/nug ...
```
`Claims.md` should just be content in a `bitrix24.md` file in the `docs` folder that follows the format used for the existing documentation: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/tree/dev/docs

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationOptions.cs:
```diff
@@ -0,0 +1,64 @@
+            ClaimActions.MapCustomJson(ClaimTypes.Name, json => string.Join(' ', json.GetProperty(UserFields.Name).GetString(), json.GetProperty(UserFields.SecondName).GetString(), json.GetProperty(UserFields.LastName).GetString()));
+        }
+
+        public string? Domain { get; set; }
```
Add an `///` comment to this property.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24PostConfigureOptions.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            if (string.IsNullOrWhiteSpace(options.Domain))
+            {
+                throw new ArgumentException("No Okta domain configured.", nameof(options));
```
```suggestion
                throw new ArgumentException("No Bitrix24 domain configured.", nameof(options));
```

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Claims.md:
```diff
@@ -0,0 +1,37 @@
+﻿Available claims
```
This should be part of a `bitrix24.md` file in the `docs` folder folding the format of the others, with some explanatory text for the claims.

martincostello commented on src/AspNet.Security.OAuth.LinkedIn/LinkedInAuthenticationDefaults.cs:
```diff
@@ -12,7 +12,7 @@ namespace AspNet.Security.OAuth.LinkedIn
     /// Default values used by the LinkedIn authentication middleware.
     /// </summary>
-    public class LinkedInAuthenticationDefaults
+    public static class LinkedInAuthenticationDefaults
```
This is a binary breaking change - please revert.

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationDefaults.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.AuthorizationEndpoint"/>.
+        /// </summary>
+        public const string AuthorizationEndpointPath = "/oauth/authorize/";
```
Just checking - correct that this ends with a `/`?

xperiandri replied:
Yes, it must. Otherwise it returns redirected and authentication fails

martincostello commented on src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationDefaults.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <summary>
+        /// Default value for <see cref="OAuthOptions.TokenEndpoint"/>.
+        /// </summary>
+        public const string TokenEndpointPath = "/oauth/token/";
```
Same here.

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/Bitrix24/Bitrix24Tests.cs:
```diff
@@ -0,0 +1,50 @@
+        {
+            builder.AddBitrix24(options =>
+            {
+                options.Domain = "Test.bitrix24.com";
```
```suggestion
                options.Domain = "test.bitrix24.local";
```

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/Bitrix24/bundle.json:
```diff
@@ -0,0 +1,58 @@
+  "$schema": "https://raw.githubusercontent.com/justeat/httpclient-interception/master/src/HttpClientInterception/Bundles/http-request-bundle-schema.json",
+  "items": [
+    {
+      "uri": "https://Test.bitrix24.com/oauth/token/",
```
```suggestion
      "uri": "https://test.bitrix24.local/oauth/token/",
```

martincostello commented on test/AspNet.Security.OAuth.Providers.Tests/Bitrix24/bundle.json:
```diff
@@ -0,0 +1,58 @@
+      }
+    },
+    {
+      "uri": "https://Test.bitrix24.com/rest/user.current.json",
```
```suggestion
      "uri": "https://test.bitrix24.local/rest/user.current.json",
```

Commit:
Created Bitrix24 documentation following the format used for the existing documentation, fixed remarks about tests

Created Bitrix24.md file
Edited path to it in README.md
Added Bitrix24 to the table in docs/README.md file
Added comment to Domain property
Fixed wrong names in Bitrix24PostConfigureOptions.cs and domain in Test files

xperiandri commented:
@martincostello can we add DisposableFixer analyzer as a separate pull request?

martincostello commented on docs/bitrix24.md:
```diff
@@ -0,0 +1,75 @@
+services.AddAuthentication(options => /* Auth configuration */)
+        .AddBitrix24(options =>
+        {
+            options.Domain = "my-domain";
```
Makes it slightly more obvious by using a more realistic-looking value.

```suggestion
            options.Domain = "my-domain.com";
```

martincostello commented on docs/bitrix24.md:
````diff
@@ -0,0 +1,75 @@
+
+#### Available claims
+
+``` C#
````
```suggestion
```csharp
```

martincostello commented on docs/bitrix24.md:
```diff
@@ -0,0 +1,75 @@
+
+### ClaimActions
+The fields of the user's profile to add as claims.
+#### Default claims (already included)
```
```suggestion

#### Default Claims
```

martincostello commented on docs/bitrix24.md:
```diff
@@ -0,0 +1,75 @@
+
+See [AspNet.Security.OAuth.Bitrix24.Bitrix24AuthenticationConstants.Scopes](/src/AspNet.Security.OAuth.Bitrix24/Bitrix24AuthenticationConstants.cs)
+
+### ClaimActions
+The fields of the user's profile to add as claims.
```
```suggestion
### Claim Actions

The fields of the user's profile to add as claims.
```

martincostello commented on docs/bitrix24.md:
````diff
@@ -0,0 +1,75 @@
+ClaimActions.MapCustomJson(ClaimTypes.Name, json => string.Join(' ', json.GetProperty(UserFields.Name).GetString(), json.GetProperty(UserFields.SecondName).GetString(), json.GetProperty(UserFields.LastName).GetString()));
+```
+
+#### Available claims
````
```suggestion
#### Available Claims
```

martincostello commented:
Closing due to lack of activity and merge conflicts - please open a new PR if you'd like to contribute this provider.
